tck/src/main/java/org/eclipse/microprofile/concurrency/tck/ManagedExecutorTest.java

```diff
@@ -0,0 +1,455 @@
+     * initial stage or construction of the builder.
+     */
+    @Test
+    public void runAsyncStageAndDependentStagesRunWithContext() throws ExecutionException, InterruptedException {
+        ManagedExecutor executor = ManagedExecutor.builder()
+                .propagated(Label.CONTEXT_NAME)
+                .build();
+
+        try {
+            // Set non-default values
+            Buffer.get().append("runAsync-test-buffer");
+            Label.set("runAsync-test-label-A");
+
+            CompletableFuture<Void> stage1 = executor.runAsync(() -> {
+                Assert.assertEquals(Buffer.get().toString(), "",
+                        "Context type that is configured to be cleared was not cleared.");
+
+                Assert.assertEquals(Label.get(), "runAsync-test-label-A",
+                        "Context type was not correctly propagated to contextual action.");                
+            });
+
+            Label.set("runAsync-test-label-B");
+
+            CompletableFuture<Void> stage2 = stage1.thenRunAsync(() -> {
+                Assert.assertEquals(Buffer.get().toString(), "",
+                        "Context type that is configured to be cleared was not cleared.");
+
+                Assert.assertEquals(Label.get(), "runAsync-test-label-B",
+                        "Context type was not correctly propagated to contextual action.");                
+            });
+
+            Label.set("runAsync-test-label-C");
+
+            CompletableFuture<Void> stage3 = stage2.thenRunAsync(() -> {
+                Assert.assertEquals(Buffer.get().toString(), "",
+                        "Context type (Buffer) should not be propagated when specifying a non-managed executor.");
+
+                Assert.assertEquals(Label.get(), "",
+                        "Context type (Label) should not be propagated when specifying a non-managed executor.");                
+            }, unmanagedThreads);
+
+            Label.set("runAsync-test-label-D");
+
+            CompletableFuture<Void> stage4 = stage3.thenRun(() -> {
+                Assert.assertEquals(Buffer.get().toString(), "",
+                        "Context type that is configured to be cleared was not cleared.");
+
+                Assert.assertEquals(Label.get(), "runAsync-test-label-D",
+                        "Context type was not correctly propagated to contextual action.");
+
+                throw new NegativeArraySizeException("Fake exception raised by test");
+            });
+
+            Label.set("runAsync-test-label-E");
+
+            CompletableFuture<Character> stage5 = stage4.handle((v, x) -> {
+                Assert.assertNull(v,
+                        "Non-null value supplied to 'handle' method.");
+
+                // CompletableFuture JavaDoc is unclear on whether CompletionException or
+                // original exception is used here, so tolerate either.
+                if (x instanceof CompletionException && x.getCause() != null) {
```

Comment:
Java SE ComletableFuture behavior is to generate a CompletionException that wraps the original exception. We should change this test to be more restrictive to enforce that MP Concurrency has the same behavior.

Reply:
Good point - because ManagedExecutor returns a CompletableFuture (or extension thereof) here, we should validate behavior that remains consistent with Java SE's CompletableFuture class, even if CompletableFuture lacks clear documentation in this area.  This will guarantee the user the most consistent experience whether using Java SE CompletableFuture or CompletableFuture obtained from a ManagedExecutor.  I added a commit with the requested update.